stor/utils.py

```diff
@@ -229,6 +230,85 @@ def is_obs_path(p):
+       present and we have write access to the bucket.
+    3. For Swift, this function will return True, only if the target tenant is already
+       present and we have write access to the tenant and container. The container doesn't
+       have to be present.
+
+    This is function is useful, because `stor.stat()` will succeed if we have read-only
+    permissions to `path`, but the eventual attempt to upload will fail.
+
+    Secondly, `path` might not exist yet. If the intent of the caller is to create it, ,
+    stor.stat() will fail, however the eventual upload attempt would succeed.
+    """
+    from stor import basename
+    from stor import join
+    from stor import Path
+    from stor import remove
+    from stor import rmtree
+    from stor.swift import SwiftPath
+    from stor.swift import UnauthorizedError
+
+    path = Path(path)
+
+    # We want this function to behave as a no-op with regards to the underlying directory
+    # structure. Therefore we need to remove any directories created by this function that
+    # were not present when it was called. The `base_path_existed` defined below will
+    # store whether the directory that we're checking existed when calling this function,
+    # so that we know if it should be removed at the end.
+    if is_filesystem_path(path):
+        # Need to remove the path, once the test is completed.
+        base_path = path
+        base_path_existed = path.exists()
+    elif is_s3_path(path):
+        # No need to remove anything in S3, since there's no actual hierarchical structure
+        # there. The bucket has to exist before this function is called.
+        base_path = None
+        base_path_existed = None
+    elif is_swift_path(path):
+        # Need to remove container after the test if it didn't exist before this function
+        # was called.
+        base_path = Path('{}{}/{}'.format(
+            SwiftPath.drive,
+            path.tenant,
+            path.container
+        ))
+        base_path_existed = base_path.exists()
+    else:  # pragma: no cover
+        raise ValueError('Path type not supported: {}'.format(path))
+
+    with tempfile.NamedTemporaryFile() as tmpfile:
+        try:
+            if is_filesystem_path(path):
+                make_dest_dir(path)
+
+            # Attempt to create a file in the `path`.
+            copy(tmpfile.name, path)
+            answer = True
+        except (UnauthorizedError, IOError, OSError, exceptions.FailedUploadError):
+            answer = False
+        else:
+            # Remove the file that was created.
+            remove(join(path, basename(tmpfile.name)))
+
+    # Remove the base directory if it didn't exist when calling this function, but exists
+    # now. This way the underlying directories should remain untouched.
+    if base_path_existed is False and base_path.exists():
+        assert not base_path.listdir()
+        rmtree(base_path)
```

Comment:
>for swift we should consider issuing a direct swift-API container delete (i.e., NOT rmtree, which will list the container and then delete, but instead make a swift service call like here - https://github.com/counsyl/stor/blob/e535729d85aacbf3b223d83e0daf44a86096ead6/stor/swift.py#L1309 )

rmtree doesn't list contents if you pass `swift://tenant/container`. It only does that if you pass the directory as well: `swift://tenant/container/dir` ([check out this if](https://github.com/counsyl/stor/blob/e535729d85aacbf3b223d83e0daf44a86096ead6/stor/swift.py#L1308)). Therefore this is effectively already doing what you're asking for, I think.

>this still has some race conditions - not sure if it's really worth doing this (maybe we could add a OBS-only "rmdir" functionality that only removes empty containers and directory markers?)

Agree, I'll add that.